src/OpenTelemetry/Instrumentation/DiagnosticSourceListener.cs

```diff
@@ -39,6 +39,11 @@ public void OnError(Exception error)
 
         public void OnNext(KeyValuePair<string, object> value)
         {
+            if (Sdk.SuppressInstrumentation)
```

Comment:
Probably a good optimization to have this, but I don't think it's strictly required. For "legacy" Activity flows we pass them through an ActivitySource for sampling. So the logic below should also catch these?

Reply:
@CodeBlanch Hmm 🤔, maybe I need you to point me to some code... because without this check here, this test I wrote would fail: 

https://github.com/open-telemetry/opentelemetry-dotnet/pull/1079/files#diff-a266e12581b9c9ae7bd148bbf91dabb8R37-R59

So, it would seem that for legacy activities, we'll need a check somewhere in addition to the check I have in the ActivityListener for the new activities.

Reply:
~You're right, my bad.~ That was a test and you passed, great job!

It looks like ActivitySourceAdapter has its own sampler, so we'll need to do similar in here:

https://github.com/open-telemetry/opentelemetry-dotnet/blob/44b55d476e9b62fb767437c9ec94c7b88655b8e7/src/OpenTelemetry/Trace/ActivitySourceAdapter.cs#L85-L131

Or maybe we can refactor so more is shared?

Reply:
> That was a test and you passed, great job!

Ha! Phew!

> It looks like ActivitySourceAdapter has its own sampler, so we'll need to do similar in here

I don't think so because this is actually downstream of the check I have in place.

So, `DiagnosticSourceListener.OnNext` --- invokes ---> `ListenerHandler.On[Start|End|Exception|Custom]` 

It is our `ListenerHandler`s that hold an instance of the `ActivitySourceAdapter` which is then the thing that simulates Start/Stop/etc like the standard ActivitySource.

Reply:
Are you saying it wouldn't work, or just that this spot is sooner? If the later, I suppose that is compelling enough a reason to keep it here and I'm good with it 👍

Reply:
It would work for preventing `this.activityProcessor.OnStart(activity);`, but the reason I put the check where I did in `DiagnosticSourceListener.OnNext` is that the check there prevents all instrumentation callbacks defined on the `ListenerHandler` from being invoked. For example, if some library had:

```csharp
myDiagnosticListener.Write("MyCustomEvent", payload);
```

If we had a `ListenerHander` wired up to this library then this `Write` would result in invoking the `OnCustom` callback. I figured it would be good to suppress _all_ aspects of the library's instrumentation.

Reply:
Makes sense good call!